### PR TITLE
Add CI variant to test the build with various compile-time flags toggled

### PR DESCRIPTION
**Pull request by ggreenway**
Signed-off-by: Greg Greenway <ggreenway@apple.com>

*Risk Level*: Low
*Testing*: Ran this build with a source change that broke the non-grpc build; verified build failed.

### CODE REVIEW

File: ci/do_ci.sh [hunk 1]
```diff
@@ -148,6 +148,19 @@ elif [[ "$1" == "bazel.dev" ]]; then
...
+elif [[ "$1" == "bazel.compiletimeoptions" ]]; then
+  # Right now, none of the available compile-time options conflict with each other. If this
+  # changes, this build type may need to be broken up.
+  COMPILE_TIME_OPTIONS="--define=signal_trace=disabled --define hot_restart=disabled --define google_grpc=disabled"
+  setup_clang_toolchain
+  # This doesn't go into CI but is available for developer convenience.
+  echo "bazel with different compiletime options build with tests..."
+  cd "${ENVOY_CI_DIR}"
+  echo "Building..."
+  bazel build ${BAZEL_BUILD_OPTIONS} ${COMPILE_TIME_OPTIONS} -c fastbuild //source/exe:envoy-static
```
Comment (PiotrSikora) [inline ci/do_ci.sh]: Could we use `-c dbg` here, so that we have at least some coverage for it (see #2566)?

File: ci/do_ci.sh [hunk 2]
```diff
@@ -148,6 +148,19 @@ elif [[ "$1" == "bazel.dev" ]]; then
   echo "Building and testing..."
   bazel test ${BAZEL_TEST_OPTIONS} -c fastbuild //test/...
   exit 0
+elif [[ "$1" == "bazel.compiletimeoptions" ]]; then
+  # Right now, none of the available compile-time options conflict with each other. If this
+  # changes, this build type may need to be broken up.
+  COMPILE_TIME_OPTIONS="--define=signal_trace=disabled --define hot_restart=disabled --define google_grpc=disabled"
```
Comment (PiotrSikora) [inline ci/do_ci.sh]: Add `--define boringssl=fips`? :)

Comment (ggreenway) [reply ci/do_ci.sh]: Yes, I'll add that as soon as the change is merged.

File: ci/do_ci.sh (hunk 1 above)
Comment (ggreenway) [reply ci/do_ci.sh]: Good idea

File: ci/do_ci.sh (hunk 2 above)
Comment (ggreenway) [reply ci/do_ci.sh]: Actually, whichever one is ready to merge first, let's put that in. Then the 2nd one can pickup this change.

Comment (PiotrSikora) [reply ci/do_ci.sh]: We need to bump the build image after FIPS is merged.

But yeah, let's not block this on it.

Comment (PiotrSikora) [reply ci/do_ci.sh]: Now that it's merged, please add it.

**Comment by alyssawilk** [comment]
Are we going to want to up our Circle resources if we're adding to CI?

**Comment by ggreenway** [comment]
> Are we going to want to up our Circle resources if we're adding to CI?

I'm not sure. Given how many builds we already have, this shouldn't significantly slow things down. And if it does, we can always disable this until we can get more capacity.

**Comment by ggreenway** [comment]
> I'm not a fan of compiletimeoptions name (maybe at least add underscores?), but LGTM otherwise.

Looking at it, underscores would be consistent with other names, like `bazel.release.server_only`.

**Comment by alyssawilk** [comment]
2 failures overnight with the same signature

https://circleci.com/gh/envoyproxy/envoy/136545?utm_campaign=workflow-failed&utm_medium=email&utm_source=notification

**Comment by PiotrSikora** [comment]
@alyssawilk this was fixed yesterday in #5356.

**Comment by alyssawilk** [comment]
Haha, sorry, CI failures go straight to inbox (for when we break the build) so I hit them before reviews.
Thanks for the quick fix!

**Pull request merged**